src/AssembleNodalGradPOpenBoundaryAlgorithm.C

```diff
@@ -0,0 +1,178 @@
+
+    b.parent_topology(stk::topology::ELEM_RANK, topoVec);
+    MasterElement& meSCS = *MasterElementRepo::get_surface_master_element(topoVec[0]);
+
+    const int nodesPerFace = meFC.nodesPerElement_;
+    const int numFaceScsIp = meFC.numIntPoints_;
+    const int nodesPerElement = meSCS.nodesPerElement_;
+    const int numElemScsIp = meSCS.numIntPoints_;
+
+    const int *ipNodeMap = meFC.ipNodeMap();
+
+    ws_face_pressure.resize(nodesPerFace);
+    double *p_facePressure = ws_face_pressure.data();
+
+    ws_elem_pressure.resize(nodesPerElement);
+    double *p_elemPressure = ws_elem_pressure.data();
+
+    ws_face_shape_function.resize(numFaceScsIp*nodesPerFace);
+    double *p_face_shape_function = ws_face_shape_function.data();
+
+    ws_elem_shape_function.resize(nodesPerElement*numElemScsIp);
+    double *p_elem_shape_function = ws_elem_shape_function.data();
+
+    if ( useShifted_ ) {
+      meFC.shifted_shape_fcn(p_face_shape_function);
+      meSCS.shifted_shape_fcn(p_elem_shape_function);
+    }
+    else {
+      meFC.shape_fcn(p_face_shape_function);
+      meSCS.shape_fcn(p_elem_shape_function);
+    }
+
+    for ( size_t k = 0 ; k < length ; ++k ) {
+      const double * areaVec = stk::mesh::field_data(exposedAreaVec, b, k);
+
+      stk::mesh::Entity const * face_node_rels = b.begin_nodes(k);
+      for ( int n = 0; n < nodesPerFace; ++n ) {
+        p_facePressure[n] = *stk::mesh::field_data(exposedPressureField, face_node_rels[n]);
+      }
+
+      const stk::mesh::Entity elem = bulk.begin_elements(b[k])[0];
+      const auto* elem_node_rels = bulk.begin_nodes(elem);
+      for (int n = 0; n < nodesPerElement; ++n) {
+        const stk::mesh::Entity node = elem_node_rels[n];
+        p_elemPressure[n] = *stk::mesh::field_data(pressureField, node);
+      }
+
+      const int faceOrdinal = bulk.begin_element_ordinals(b[k])[0];
+      for (int ip = 0; ip < numFaceScsIp; ++ip) {
+        const auto* areav = &areaVec[ip * nDim];
+        const int nn = ipNodeMap[ip];
+
+        stk::mesh::Entity nodeNN = face_node_rels[nn];
+        double *gradQNN = stk::mesh::field_data(GpField, nodeNN);
+        const double volNN = *stk::mesh::field_data(dualNodalVolume, nodeNN);
+
+        double pIp = 0.0;
+        const int offSet = ip * nodesPerFace;
+        for (int ic = 0; ic < nodesPerFace; ++ic) {
+          pIp += p_face_shape_function[offSet + ic] * p_facePressure[ic];
+        }
+
+        // evaluate pressure at opposing face.  If zeroGrad_ option is used, then
+        // we'll copy this value to the exposed face.  Otherwise, it's unused
+        double pOpp = 0.0;
+        const int opp_offset = meSCS.opposingFace(faceOrdinal, ip) * nodesPerElement;
+        const double* oppShapeFns = &p_elem_shape_function[opp_offset];
+        for (int ic = 0; ic < nodesPerElement; ++ic) {
+          pOpp += oppShapeFns[ic] * p_elemPressure[ic];
+        }
+
+        double press_div_vol = (zeroGrad_ ? pOpp : pIp) / volNN;
```

Comment:
@rcknaus Can you explain reason to evaluate both and use a ternary to select the appropriate value? Is this in preparation for NGP?

Reply:
Not really.  It's more following the style of masking out these options, like we do for NOC or divU etc